gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/datafetcher/EntityInteractionsFetcher.java

```diff
@@ -191,6 +281,72 @@ private EntityInteractionQueryResponse executeQueryRequest(
     return new EntityInteractionQueryResponse(entityInteractionQueryRequest, resultSet);
   }
 
+  private List<EntityInteractionQueryRequest> prepareInteractionIdsQueryRequest(
+      RequestContext requestContext,
+      EntitiesRequest request,
+      InteractionsRequest interactionsRequest,
+      boolean incoming) {
+    Set<String> otherEntityTypes = getOtherEntityTypes(interactionsRequest.getFilter());
+    if (otherEntityTypes.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<Filter> childFilters = new ArrayList<>();
+    childFilters.add(
+        QueryRequestUtil.createBetweenTimesFilter(
+            AttributeMetadataUtil.getTimestampAttributeId(metadataProvider, requestContext, SCOPE),
+            request.getStartTimeMillis(),
+            request.getEndTimeMillis()));
+
+    this.buildSpaceQueryFilterIfNeeded(requestContext, request.getSpaceId())
+        .ifPresent(childFilters::add);
+
+    List<String> idColumns =
+        getEntityIdColumnsFromInteraction(
+            DomainEntityType.valueOf(request.getEntityType()), !incoming);
+
+    // Group by the entity id column first, then the other end entity type for the interaction.
+    List<org.hypertrace.core.query.service.api.Expression> idExpressions =
+        idColumns.stream()
+            .map(QueryRequestUtil::createAttributeExpression)
+            .collect(Collectors.toUnmodifiableList());
+
+    // adding empty selections
+    List<org.hypertrace.core.query.service.api.Expression> selections = new ArrayList<>();
+    selections.add(
+        QueryRequestUtil.createCountByColumnSelection(
+            Optional.ofNullable(idColumns.get(0)).orElseThrow()));
+
+    Map<String, QueryRequest> queryRequests = new HashMap<>();
```

Comment:
If I'm reading how the map is used, we're doing a for loop, putting in the map then transforming the map at the end. Instead, stream the entity types and collect requests.